### PR TITLE
Add scripts

### PR DESCRIPTION
**Pull request by HarrisonGregg**
Adds some scripts for data preparation and interacting with the classifiers.  Also adds some features to the classifiers that are used in the scripts.

### CODE REVIEW

File: osp_classifier/classifiers.py [hunk 1]
```diff
@@ -111,3 +123,47 @@ def __init__(self):
...
+                norm='l1'
+            )),
+            ('clf', SGDClassifier(
+                max_iter=1000
+            )),
+        ])
+
+    def prep(self, X):
+        X = X[X['Text'].map(len) > 150]
+        X = X[X['PrimaryField'].map(bool)]
```
Comment (jodizzle) [inline osp_classifier/classifiers.py]: Sorry, what does this do?

File: bin/cli.py [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+
+logger = logging.getLogger("osp_classifier")
+
+
+@click.command()
+@click.argument("command")
+@click.argument("classifier")
+@click.argument("csv_in", type=click.Path(exists=True))
+@click.argument("csv_out", required=False)
+@click.option("--chunksize", type=int, default=sys.maxsize
```
[the remark this answers is not in the document]
Comment (HarrisonGregg) [reply bin/cli.py]: Fixed

File: bin/cli.py [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+              help="How many rows of the input CSV to process at a time")
+@click.option("--dataset", help="Only use rows with this dataset name")
+@click.option("--not_dataset", help="Exclude rows with this dataset name")
```
[the remark this answers is not in the document]
Comment (HarrisonGregg) [reply bin/cli.py]: `df = df[df['Dataset'].eq([dataset1, dataset2])]` actually does an elementwise comparison, but I found another way to do it.

File: bin/cli.py [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+@click.option("--output_filter", help="Only save rows with this new value")
```
[the remark this answers is not in the document]
Comment (HarrisonGregg) [reply bin/cli.py]: As the number of arguments and options grew, it just became too long to have them all spread accross multiple lines, so I guess one line is my preferred approach now.

File: bin/cli.py [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+        elif command == "score":
+            clf = Classifier.load()
+            df = clf.prep(df)
+            score = clf.score(df, df[Classifier.result_column])
+            logger.info(f"Accuracy is {score}")
+
+        elif command == "apply":
+            clf = Classifier.load()
+            df[Classifier.result_column] = clf.predict(df)
+            write_mode = "a" if i else "w"
```
[the remark this answers is not in the document]
Comment (HarrisonGregg) [reply bin/cli.py]: The downside of that approach is that it would break up the logic used for `apply` so that it wasn't contained in the same block, since `apply` is the only command that would need the buffer opened.  I don't think it's worth that sacrifice to avoid needing to set the write mode, but it's a good idea for other scripts.

I'm not clear on what difference the `chunksize` in `to_csv` makes, but I don't think it's intended to be used with the `chunksize` in `read_csv`.

File: bin/cli.py [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+            df = df[df[Classifier.result_column].str.eq(output_filter)]
```
[the remark this answers is not in the document]
Comment (HarrisonGregg) [reply bin/cli.py]: Fixed

File: bin/create_metadata_csv.py [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+    type=click.Path()
+)
+@click.argument(
+    "data_path",
+    type=click.Path()
+)
+def main(file_path, csv_path, data_path):
+    """Create a CSV with metadata for the WARC files listed in a text file.
+    """
+
```
[the remark this answers is not in the document]
Comment (HarrisonGregg) [reply bin/create_metadata_csv.py]: Fixed

File: bin/remove_csv_columns.py [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+    "csv_out",
+    type=click.Path()
+)
+@click.argument(
+    "columns"
+)
+@click.option(
+    "--chunksize",
+    type=int,
+    default=10000,
```
[the remark this answers is not in the document]
Comment (HarrisonGregg) [reply bin/remove_csv_columns.py]: It's only `sys.maxsize` in `cli.py` because training a classifier can't be chunked, but otherwise, it seems like chunking makes sense to do by default.

File: osp_classifier/classifiers.py [hunk 9]
```diff
@@ -3,30 +3,35 @@
...
-logger = logging.getLogger(__name__)
+from osp_classifier.transformers import ColumnExtractor, ColumnCombiner
 
 
 class ClassifierBase():
     """Abstract base class for OSP classifiers.
     """
     def __init__(self):
-        raise NotImplementedError
+        self.logger = logging.getLogger(self.__class__.__name__)
```
[the remark this answers is not in the document]
Comment (HarrisonGregg) [reply osp_classifier/classifiers.py]: In general, I don't think it's a reasonable expectation to not do anything in a base class's `__init__` method.  There might be other more important things that we'd like to initialize for all subclasses there, and it seems weird to restrict ourselves from being able to do that in `__init__` just to emphasize that the base class is abstract.

I definitely think that there's more value in having log messages indicate which classifier they are for when running in production than there is in preventing someone from instantiating the base class.

File: osp_classifier/classifiers.py [hunk 10]
```diff
@@ -3,30 +3,35 @@
...
+
+    def prep(self, X):
+        return X
 
     def fit(self, X, y=None):
+        start = time.time()
         self.pipeline.fit(X, y)
+        end = time.time()
+        self.logger.info(f"Fitting took {end-start} seconds.")
```
[the remark this answers is not in the document]
Comment (HarrisonGregg) [reply osp_classifier/classifiers.py]: I think `info` makes sense here.  I think of `debug` as rather verbose output that might be helpful in debugging, which doesn't seem to fit the use case here.

File: osp_classifier/classifiers.py [hunk 11]
```diff
@@ -75,7 +84,10 @@ def save(self, pickle_folder=pickle_path, compress=9, protocol=-1):
 class SyllabusClassifier(ClassifierBase):
     """Class for handling the pipeline used for the syllabus classifier.
     """
-    def __init__(self):
+    result_column = "IsSyllabus"
```
[the remark this answers is not in the document]
Comment (HarrisonGregg) [reply osp_classifier/classifiers.py]: Sure, that makes sense

File: osp_classifier/classifiers.py [hunk 12]
```diff
@@ -111,3 +123,47 @@ def __init__(self):
                 random_state=983214
             ))
         ])
+
+    def predict(self, X, threshold=None):
+        if not threshold:
+            return super().predict(X)
+        else:
+            return self.pipeline.predict_proba(X)[:, 1] > threshold
```
[the remark this answers is not in the document]
Comment (HarrisonGregg) [reply osp_classifier/classifiers.py]: We could, but it doesn't seem necessary for an internal method.

File: osp_classifier/classifiers.py (hunk 1 above)
Comment (HarrisonGregg) [reply osp_classifier/classifiers.py]: It filters out all the rows whose `PrimaryField` is empty.

File: bin/cli.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (jodizzle) [reply bin/cli.py]: I don't think it's that big a deal to stick something like
```python
if command == "apply":
    f = open(csv_out, "w")
```
above the loop.  In fact, it might be nice to have some extra logic up there anyway, to throw some kind of warning or error if the user defines a `csv_out` when they're not using the "apply" command.  But if you want to avoid that, I can think of two options:

- Define `csv_out` with `type=click.File("w")`, as discussed [here](http://click.pocoo.org/5/arguments/#file-arguments), and not have to worry about opening a buffer.  (I think this mode could actually help in several places throughout the scripts; sorry for not recommending it earlier.)
- Just use mode `"a"`, since that will create the file if it doesn't exist.  Then you could have some extra logic for deleting the file if it does exist, if you cared about that.

Basically, the use of enumerate here just strikes me as a bit more complicated than it needs to be.

File: bin/cli.py [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+      SyllabusClassifier: Predict whether a document is a syllabus or not.
+      FieldClassifier: Predict a primary field for a syllabus.
+
+    CSV_IN
+      The path of the CSV to load the input data from.
+
+    CSV_OUT (only required for 'apply' command)
+      The path of the CSV to save the output to.
+    """
+    Classifier = getattr(osp_classifier.classifiers, classifier, None)
```
Comment (jodizzle) [inline bin/cli.py]: Is there a reason this variable is capitalized?

File: bin/create_metadata_csv.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (jodizzle) [reply bin/create_metadata_csv.py]: Was it?

File: osp_classifier/classifiers.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (jodizzle) [reply osp_classifier/classifiers.py]: Sure, I suppose that's fine.  In that case, it might be advisable to upgrade to more formal abstract class-ness (e.g., [`abc`](https://docs.python.org/3/library/abc.html)) at some point in the future.  However, I don't think it's necessary as of right now.

File: osp_classifier/classifiers.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (jodizzle) [reply osp_classifier/classifiers.py]: Yeah, I think that's a good way of looking at it.

File: osp_classifier/classifiers.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (jodizzle) [reply osp_classifier/classifiers.py]: Well, the idea, at least for me, was that `ClassifierBase` would be a container for the `self.pipeline` attribute: it'd reroute all the pipeline methods (or at least the ones we found useful) to `self.pipeline`, with small modifications as needed, and then it'd tack on some additional methods.  Along those lines, I think it would make sense to keep parity with `sklearn.pipeline.Pipeline` and implement `predict_proba` distinctly.

That being said, this method seems like it could be useful even outside of `SyllabusClassifier`, so I think it would be reasonable to basically move the logic to `ClassifierBase`, and have `ClassifierBase.predict` look like:

```python
def predict(self, X, threshold=None):
    if not threshold:
        return self.pipeline.predict(X)
    else:
        return self.predict_proba(X)[:, 1] > threshold
```

Does that sound reasonable?

File: bin/cli.py (hunk 13 above)
Comment (HarrisonGregg) [reply bin/cli.py]: It's a class, so shouldn't it be capitalized?

File: osp_classifier/classifiers.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (HarrisonGregg) [reply osp_classifier/classifiers.py]: Oh cool.  I didn't know there was a package for that.  Yeah, it might make sense to do that.

File: osp_classifier/classifiers.py [hunk 14]
```diff
@@ -3,30 +3,35 @@
...
 
 
 class ClassifierBase():
     """Abstract base class for OSP classifiers.
     """
     def __init__(self):
-        raise NotImplementedError
+        self.logger = logging.getLogger(self.__class__.__name__)
+
+    def prep(self, X):
```
[the remark this answers is not in the document]
Comment (HarrisonGregg) [reply osp_classifier/classifiers.py]: I changed it to `preprocess`

File: bin/cli.py (hunk 13 above)
Comment (jodizzle) [reply bin/cli.py]: Ah, my bad.  I was confused and thought it was an instance of a class.

File: setup.py [hunk 15]
```diff
@@ -14,5 +14,14 @@
     author="OSP authors",
     author_email="syllabusopen@gmail.com",
 
-    packages=["osp_classifier"]
+    packages=["osp_classifier"],
+
+    py_modules=['osp_classifier'],
```
Comment (jodizzle) [inline setup.py]: Is this necessary when we already have `packages=` above?

File: setup.py [hunk 16]
```diff
@@ -14,5 +14,14 @@
...
+    install_requires=[
+        'Click',
+    ],
+    entry_points='''
+        [console_scripts]
+        osp_classifier=bin.cli:cli
+    ''',
```
Comment (jodizzle) [inline setup.py]: Can you really define `entry_points` like this?  It seems to work here as intended, but I was looking it up and the go-to way seems to be to have it be a dictionary: https://setuptools.readthedocs.io/en/latest/setuptools.html#automatic-script-creation

File: bin/cli.py [hunk 17]
```diff
@@ -0,0 +1,78 @@
...
+            clf = Classifier()
+            df = clf.preprocess(df)
+            clf.fit(df, df[Classifier.target_column])
+            clf.save()
+
+        elif command == "score":
+            clf = Classifier.load()
+            df = clf.preprocess(df)
+            score = clf.score(df, df[Classifier.target_column])
+            logger.info(f"Accuracy is {score}")
```
Comment (jodizzle) [inline bin/cli.py]: Since the point of this command is to print data, I think you should do a normal `print` statement rather than log.

**Comment by jodizzle** [comment]
`bin/cli.py` is missing an `if __name__ == "__main__":` statement.  Is this intentional?

**Comment by jodizzle** [comment]
Is the `bin/__init__.py` file necessary?  I think you can define an entry point to `bin/cli.py` even without it.

File: bin/test_syllabus_classifier_threshold.py [hunk 18]
```diff
@@ -0,0 +1,49 @@
+#!/usr/bin/env python3
+
+import logging
+
+import click
+import pandas as pd
+import numpy as np
+from sklearn.metrics import confusion_matrix
+
+from osp_classifier.classifiers import OSPLogisticRegressionCV
```
[the remark this answers is not in the document]
Comment (HarrisonGregg) [reply bin/test_syllabus_classifier_threshold.py]: Fixed

File: bin/cli.py (hunk 17 above)
Comment (HarrisonGregg) [reply bin/cli.py]: I'm not sure I follow your logic there.  Why would you ever use a `print` statement if you're using `logging`?  I would think that if you set the log level to `WARNING` or higher, it shouldn't print out things like this.

File: setup.py (hunk 16 above)
Comment (HarrisonGregg) [reply setup.py]: I just copied it from somewhere.  Using a dict does seem to make more sense

File: setup.py (hunk 15 above)
Comment (HarrisonGregg) [reply setup.py]: Hmm, I guess not.

File: osp_classifier/classifiers.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (HarrisonGregg) [reply osp_classifier/classifiers.py]: Unfortunately, this only works for a binomial classifier, so it wouldn't make sense to implement it in `ClassifierBase`, since that will be used for multinomial classifiers as well.

File: bin/cli.py (hunk 17 above)
Comment (jodizzle) [reply bin/cli.py]: Well, this block only gets executed if the user specifies `"score"` as the command, right?  It doesn't have any other purpose?  Given that, the score should be printed regardless of what `LOG_LEVEL` the user has set, since the printing of the score is the expected output, not supplementary information.

File: osp_classifier/classifiers.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (jodizzle) [reply osp_classifier/classifiers.py]: Well, I was thinking that if `predict_proba` wasn't available in the underlying classifier, then so be it, similar to what `sklearn.pipeline.Pipeline` does.  But the general approach of `ClassifierBase` at the moment is pretty silly, so we can leave it for now.  It's not particularly relevant to the idea of this PR anyway.

File: bin/cli.py [hunk 19]
```diff
@@ -0,0 +1,83 @@
...
+
+
+logger = logging.getLogger("osp_classifier")
+
+
+@click.command()
+@click.argument("command")
+@click.argument("classifier")
+@click.argument("csv_in", type=click.Path(exists=True))
+@click.argument("csv_out", required=False)
```
Comment (jodizzle) [inline bin/cli.py]: I think using `type=click.File()`, like in the [other script](https://github.com/opensyllabus/osp-classifier/blob/c91efd8a421c28be9b4e003fb475c7a4148860c5/bin/extract_text.py#L25), would be prudent.

File: bin/cli.py [hunk 20]
```diff
@@ -0,0 +1,83 @@
...
+            df = clf.preprocess(df)
+            score = clf.score(df, df[Classifier.target_column])
+            logger.info(f"Accuracy is {score}")
+
+        elif command == "apply":
+            clf = Classifier.load()
+            df[Classifier.target_column] = clf.predict(df)
+            if output_filter:
+                df = df[df[Classifier.target_column].str.eq(output_filter)]
+            write_mode = "a" if i else "w"
```
Comment (jodizzle) [inline bin/cli.py]: I think if you change the `type` to `click.File`, as recommended above, you can remove this check, as was [done for the other script](https://github.com/opensyllabus/osp-classifier/commit/7afcd3850570f6697d4a1da80bb5ddac74e1fb7e#diff-028c80648f2e9923424db9dd3568f8aa).

File: bin/create_metadata_csv.py [hunk 21]
```diff
@@ -0,0 +1,68 @@
...
+import pandas as pd
+
+from osp_classifier.warc_utils import parse_warc
+
+
+logger = logging.getLogger("create_metadata_csv")
+
+
+@click.command()
+@click.argument("file_in", type=click.Path(exists=True))
```
Comment (jodizzle) [inline bin/create_metadata_csv.py]: Should this be named `csv_in`, for consistency?

File: bin/create_metadata_csv.py [hunk 22]
```diff
@@ -0,0 +1,68 @@
...
+    for i, keys_df in enumerate(chunks):
+        df = keys_df[0].map(get_warc_metadata_from_key).apply(pd.Series)
+        df = df[df.any(axis=1)]
+        df.to_csv(out_file, header=not i, index=False)
+        total_written += len(df)
+        logger.info(f"{total_written} entries written to CSV")
+
+
+
+if __name__ == "__main__":
```
Comment (jodizzle) [inline bin/create_metadata_csv.py]: I think we'd want two empty lines above this, not three, right?

File: bin/create_metadata_csv.py [hunk 23]
```diff
@@ -0,0 +1,68 @@
...
+
+from osp_classifier.warc_utils import parse_warc
+
+
+logger = logging.getLogger("create_metadata_csv")
+
+
+@click.command()
+@click.argument("file_in", type=click.Path(exists=True))
+@click.argument("csv_out", type=click.Path())
```
Comment (jodizzle) [inline bin/create_metadata_csv.py]: You might consider using `click.File` as the `type` for these arguments as well.  Same for the other scripts.

**Comment by HarrisonGregg** [comment]
The intention is for `bin/cli.py` to be used with the `osp_classifier` endpoint instead of being run directly, so I didn't include `if __name__ == "__main__":`.

File: bin/cli.py (hunk 17 above)
Comment (HarrisonGregg) [reply bin/cli.py]: I suppose that makes sense.

**Comment by jodizzle** [comment]
@HarrisonGregg If you definitely want it that way, then I think you should make it so that `cli.py` isn't executable (i.e., `chmod -x bin/cli.py`) and remove `#!/usr/bin/env python3` from the top of the file.

**Comment by HarrisonGregg** [comment]
@jodizzle I think I addressed all your comments.  Want to take another look?

File: bin/create_metadata_csv.py [hunk 24]
```diff
@@ -0,0 +1,67 @@
...
+def main(file_in, csv_out, data_path, chunksize):
+    """Create a CSV with metadata for the WARC files listed in a text file.
+
+    \b
+    FILE_IN
+      The path of the file with keys separated by newlines to load the
+      input data from.
+    CSV_OUT
+      The path of the CSV to save the output to.
+    DATA
```
Comment (jodizzle) [inline bin/create_metadata_csv.py]: This should be "DATA_PATH" to match the argument name.  Or, alternatively, to match the [`extract_text.py`](https://github.com/opensyllabus/osp-classifier/blob/c91efd8a421c28be9b4e003fb475c7a4148860c5/bin/extract_text.py#L41) script, it could be "DATA_ROOT".

File: bin/cli.py [hunk 25]
```diff
@@ -0,0 +1,80 @@
...
+@click.argument("command")
+@click.argument("classifier")
+@click.argument("csv_in", type=click.File("r"))
+@click.argument("csv_out", required=False, type=click.File("w"))
+@click.option("--chunksize", type=int, default=sys.maxsize,
+              help="How many rows of the input CSV to process at a time")
+@click.option("--dataset", multiple=True,
+              help="Only use rows with given dataset name(s)")
+@click.option("--not_dataset",  multiple=True,
+              help="Exclude rows with givene dataset name(s)")
```
Comment (jodizzle) [inline bin/cli.py]: Pedantic: Misspelling.

**Comment by jodizzle** [comment]
I've addressed my two remarks, and made one other tweak.  Merging.
